src/query/functions/linear/sort.go

```diff
@@ -20,13 +20,148 @@
 
 package linear
 
-const (
-	// NB: Because Prometheus's sort and sort_desc only look at the last value,
-	// these functions are essentially noops in M3 as we don't support instant queries.
+import (
+	"fmt"
+	"sort"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+)
 
+const (
 	// SortType returns timeseries elements sorted by their values, in ascending order.
 	SortType = "sort"
 
 	// SortDescType is the same as sort, but sorts in descending order.
 	SortDescType = "sort_desc"
 )
+
+type sortOp struct {
+	opType string
+	lessFn lessFn
+}
+
+// OpType for the operator
+func (o sortOp) OpType() string {
+	return o.opType
+}
+
+// String representation
+func (o sortOp) String() string {
+	return fmt.Sprintf("type: %s", o.opType)
+}
+
+type sortNode struct {
+	op         sortOp
+	controller *transform.Controller
+}
+
+type valueAndMeta struct {
+	val        float64
+	seriesMeta block.SeriesMeta
+}
+
+type lessFn func (i, j float64) bool
+
+// Node creates an execution node
+func (o sortOp) Node(
+	controller *transform.Controller,
+	_ transform.Options,
+) transform.OpNode {
+	return &sortNode{
+		op:         o,
+		controller: controller,
+	}
+}
+
+func (n *sortNode) Params() parser.Params {
+	return n.op
+}
+
+func (n *sortNode) Process(queryCtx *models.QueryContext, ID parser.NodeID, b block.Block) error {
+	return transform.ProcessSimpleBlock(n, n.controller, queryCtx, ID, b)
+}
+
+func (n *sortNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID, b block.Block) (block.Block, error) {
+	if !queryCtx.Options.Instantaneous {
+		return b, nil
+	}
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	meta := b.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	return n.processInstantBlock(queryCtx, stepIter, meta, seriesMetas)
+}
+
+func (n *sortNode) processInstantBlock(queryCtx *models.QueryContext, stepIter block.StepIter, meta block.Metadata, seriesMetas []block.SeriesMeta) (block.Block, error) {
+	ixLastStep := stepIter.StepCount() - 1 //we only care for the last step values for the instant query
+	for i := 0; i <= ixLastStep; i++ {
+		if !stepIter.Next() {
+			return nil, fmt.Errorf("invalid step count; expected %d got %d", stepIter.StepCount(), i+1)
+		}
+	}
+	values := stepIter.Current().Values()
+	meta.ResultMetadata.KeepNaNs = true
+	valuesToSort := make([]valueAndMeta, len(values))
+	for i, value := range values {
+		valuesToSort[i] = valueAndMeta{
+			val:        value,
+			seriesMeta: seriesMetas[i],
+		}
+	}
```

Comment:
nit: safer to use append here

Reply:
+1

Reply:
We are relying on index when setting from `seriesMetas[i]` so `append` would just hide this and make it less clear.

Reply:
Yeah, it's just we've seen issues before with forward compat where future changes to slice size caused out of index errors, while append will deal with this case